Changes to encryption wrapper

sharidas opened the pull request:
Changes made to encryption wrapper so that we can
get the commands like transfer-ownership or recreate
masterkey work. This change doesn't alter the core
functionality of copy or fopen function. The arguments
passed to the function remains same as other wrappers.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
<!--- Describe your changes in detail -->
Changes made to the ecnryption warpper and view to get the transfer-ownership command recreate master key work. Basically there was a problem which caused to change the arguments passed to the routines like fopen or copy. With this change we don't have to alter them. But we use flags to control the code flow for the commands.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
This change would prevent the modification to common routines which were made to arguments of functions like fopen or copy etc.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- [x] Scratch installation of oC and create 2 users
- [x] Verified transfer ownership command works with and without encryption between two users.
- [x] Verified Recreate command for master key works on the same setup.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

sharidas committed:
Changes to encryption wrapper

Changes made to encryption wrapper so that we can
get the commands like transfer-ownership or recreate
masterkey work. This change doesn't alter the core
functionality of copy or fopen function. The arguments
passed to the function remains same as other wrappers.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -260,7 +270,18 @@ protected function transfer(OutputInterface $output) {
+		 * If encryption is enabled and masterkey is the option selected
+		 * kindly use the CustomView wrapper.
+		 */
+		/*if ($this->encryptionManager->isEnabled() &&
```
please remove

PVince81 commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -86,6 +86,10 @@ class Encryption extends Wrapper {
 
+	private $sourcePath;
+
+	private $decryptedFile = false;
```
so if I understand correctly, this flag only affects write operations ?

If yes, call it `$disableWriteEncryption` instead and the methods `setDisableWriteEncryption()`

PVince81 commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -622,6 +644,22 @@ public function moveFromStorage(Storage $sourceStorage, $sourceInternalPath, $ta
+	 * Set the flag to true, so that the file would be
+	 * in the decrypted state.
+	 */
+	public function setDecryptedFileFlag() {
```
Just pass the flag as argument, no need to write two methods...

PVince81 commented on lib/private/Files/View.php:
```diff
@@ -2124,4 +2145,14 @@ private function createParentDirectories($filePath) {
+	 * the storage's set value for decrypted file flag is called.
+	 */
+
+	public function setDecryptedFile() {
```
I'd prefer that the View doesn't know anything about encryption. Encryption is just like a plugin that inserts itself so it's not good that the core code like View relies on encryption specific things.

Let's make flag static in the Encryption class above instead and set it from outside. That should do for now.

PVince81 reviewed (commented):
great progress, see comments

let's use the static approach

PVince81 commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -451,14 +456,31 @@ public function fopen($path, $mode, $sourceFileOfRename = null) {
-					$size, $unencryptedSize, $headerSize, $signed, $sourceFileOfRename);
+					$size, $unencryptedSize, $headerSize, $signed, $this->sourcePath[$path]);
+
+				if (isset($this->sourcePath[$path])) {
```
you need to call isset earlier, else line 478 is going to throw a warning when not set

PVince81 replied:
@sharidas did you miss this comment ?

sharidas replied:
No, I thought in a different way. `wrap()` expects either null or string `$this->sourcePath[$path]`. So if its not set then I set as null at https://github.com/owncloud/core/pull/28774/files#diff-938b56891f3079e628cfda1c331911b7R476 .
Else if its set then unset at line 484. So is there something really wrong here?

PVince81 commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -748,10 +786,11 @@ private function copyBetweenStorage(Storage $sourceStorage, $sourceInternalPath,
+					$this->sourcePath[$targetInternalPath] = $absSourcePath;
 				} else {
-					$target = $this->fopen($targetInternalPath, 'w');
+					$this->sourcePath[$targetInternalPath] = null;
```
use `unset` for maximum deletion

sharidas commented:
- [x] Removed changes from View which points to encryption. 
- [x] Removed the comment ( code comment ) 
- [x] Renamed the variable `$decryptedFile` to `$disableWriteEncryption` and also renamed the function associated with the operation to `setDisableWriteEncryption()`. Argument is passed to `setDisableWriteEncryption()` for operation set or unset.
- [x]

PVince81 commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -451,14 +456,35 @@ public function fopen($path, $mode, $sourceFileOfRename = null) {
+				 * for write mode.
+				 */
+				if (self::$disableWriteEncryption && ($mode !== 'r')) {
+					return fopen($this->storage->getSourcePath($path), $mode);
```
This is going to break horribly with external storage because you are doing a fopen directly on the filesystem.

Use `$this->getWrapperStorage()->fopen()` instead

sharidas replied:
@PVince81 Updated in my latest PR. Thanks for pointing it out.

sharidas commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -451,14 +456,34 @@ public function fopen($path, $mode, $sourceFileOfRename = null) {
-					$this->uid, $encryptionModule, $this->storage, $this, $this->util, $this->fileHelper, $mode,
-					$size, $unencryptedSize, $headerSize, $signed, $sourceFileOfRename);
+
+				if (isset($this->sourcePath[$path])) {
```
Hope this looks better.

PVince81 commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -622,6 +647,15 @@ public function moveFromStorage(Storage $sourceStorage, $sourceInternalPath, $ta
+	 *
+	 * @param $isDisabled bool
+	 */
+	public function setDisableWriteEncryption($isDisabled) {
```
function must be static

sharidas replied:
Done.

sharidas commented:
- [x] Made the function setDisableWriteEncryption staic.
- [x] Used a local variable to pass the `$this->sourcePath[$path]` if its set else null to -> `wrap()`
- [x] Removed the view code

PVince81 commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -86,6 +86,10 @@ class Encryption extends Wrapper {
 	/** @var  ArrayCache */
 	private $arrayCache;
 
+	private $sourcePath;
```
PHPDoc comments please

PVince81 reviewed (approved):
:+1: add missing PHP doc then this is good to go

PVince81 commented:
@sharidas please backport to stable10

sharidas commented:
Backport of this PR: https://github.com/owncloud/core/pull/28845
